Coffee shop

eandreea opened the pull request:
Mersi de hint :), tocmai partea de fork nu am citit-o cand m-am uitat la partea de pull request.
Partea de app a ramas in starea de ieri, nu ruleaza conform cerintelor dupa cum v-am explicat in mail:  " am facut totul intr-un playground, dar surpriza a aparut cand am vrut sa rulez playgroundul la sfarsit cu tot cu partea de teste si playground support in el, cand am primit error: cannot load underlying module for 'XCTest' "
Aici trebuie sa vad ce solutii exista.

eandreea committed:
initial playground

eandreea committed:
test_applicationValidArguments fail

eandreea committed:
test_applicationValidArguments success

eandreea committed:
location test fail

eandreea committed:
location test success

eandreea committed:
test_applicationValidArguments user location and shops data file fail

eandreea committed:
test_applicationValidArguments user location and shops data file success

eandreea committed:
test valid filename fail

eandreea committed:
test valid filename success

eandreea committed:
test new coffee shop fail

eandreea committed:
test new coffee shop success

eandreea committed:
test coffee shop from string fail

eandreea committed:
test coffee shop from string success

eandreea committed:
test coffee shops from file content string fail

eandreea committed:
test coffee shops from file content string success

eandreea committed:
test coffee shops ordered by distance to given location fail

eandreea committed:
test distance to location fail

eandreea committed:
test coffee shops ordered by distance to given location success

eandreea committed:
3 closest coffeeshops

eandreea committed:
Read file from Shared Playground Data folder in user Documents directory (terminal & playground)

eandreea committed:
comment unit test code to run from command line

alexandru-calinoiu commented:
Please add minimum instructions on how to run the code, like the expected location for the file

eandreea committed:
How to test the application

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -1,6 +1,18 @@
 
+struct CoffeeShopApp {
+
+    init?(_ arguments: [String]) {
```
Way to much code for the first test to pass.

You actually want to add as little code as possible to make the test pass, in this case parameter validation/parsing was not in the scope

eandreea replied:
Ok, will refactor it. Thought that the arguments should be mandatory for the app and should be validated

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -30,3 +30,28 @@ class CoffeeShopAppTest: XCTestCase {
+    override func tearDown() {
+    }
+
+    func test_defaultLocation() {
```
One should only write one test that fails.

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -1,6 +1,18 @@
 
+struct CoffeeShopApp {
+
+    init?(_ arguments: [String]) {
```
Why use `init?`?

eandreea replied:
The coffee shop app object should not initialise if the provided command line arguments are not as required (x, y, file_name). To run a coffee shop app a user location and coffee shop data file are required.

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -31,6 +31,11 @@ class CoffeeShopAppTest: XCTestCase {
 CoffeeShopAppTest.defaultTestSuite.run()
 
+struct Location {
+    var x = 0.0
```
Why do we want to have the location mutable? Is generally not a good idea to have this value objects muttable

eandreea replied:
Yes, you are right

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -1,65 +1,91 @@
+// MARK: - CoffeeShopApp
 struct CoffeeShopApp {
 
+    var user = Location()
```
Why the variable name is user but the type is Location?

eandreea replied:
Should have named it userLocation

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -42,14 +42,26 @@ struct CoffeeShopApp {
             let y = Double(arguments[0]),
-            let x = Double(arguments[1])
+            let x = Double(arguments[1]),
+            CoffeeShopApp.isFilename(arguments[2])
```
This is not tested. Saw tests only for `isFilename`.
What is this call used for ?

eandreea replied:
test_applicationNotEnoughArguments, test_applicationYNotDouble and test_applicationXNotDouble test the failable initializer

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -58,6 +58,39 @@ class CoffeeShopTest: XCTestCase {
     }
+
+    func test_coffeeShopFromString() {
+        let coffeeShop = CoffeeShop("Starbucks Seattle,47.5809,-122.3160", separator: ",")
```
Would prefer to have the constructor as dumb as possible. 

https://tamasgyorfi.net/2016/04/16/work-in-the-constructor-anti-pattern/

eandreea replied:
I've made it as a convenience initializer for when we read the csv file and basically each line represents a coffee shop.

eandreea replied:
This is what I had in mind when i used the convenience initializer this way: "Create convenience initializers whenever a shortcut to a common initialization pattern will save time or make initialization of the class clearer in intent."
https://docs.swift.org/swift-book/LanguageGuide/Initialization.html

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -34,7 +34,7 @@ LocationTest.defaultTestSuite.run()
 
 // MARK: - CoffeeShop
-struct CoffeeShop {
+struct CoffeeShop: Equatable {
```
Why is this needed?

eandreea replied:
In order to compare 2 instances of coffee shop, the class needs to conform to the Equatable protocol.

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -166,6 +166,30 @@ class CoffeeShopTest: XCTestCase {
+        XCTAssertNotNil(coffeeShops)
+        XCTAssertEqual(coffeeShops!.count, 6)
+        let location = Location(x: -122.4, y: 47.6)
+        let orderedCoffeeShops = CoffeeShop.sort(coffeeShops: coffeeShops, byDistanceTo: location, ascending: true)
```
This looks more like a procedural way of doing things, in which data is not grouped together with the logic that can act on it.

eandreea replied:
Yes

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -69,6 +78,17 @@ struct CoffeeShop: Equatable {
+        var coffeeShopWithDistance = coffeeShops.map { CoffeeShopWithDistance(coffeeShop: $0, distance: $0.location.distanceTo(location)) }
+        coffeeShopWithDistance.sort {
+            if ascending {
+                return $0.distance < $1.distance
```
What is the benefit of using $0 / $1 over named parameters ?

eandreea replied:
I've used swift shorthand argument names https://docs.swift.org/swift-book/LanguageGuide/Closures.html#ID97

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -26,6 +26,9 @@ class CoffeeShopAppTest: XCTestCase {
         XCTAssertNotNil(app)
+        XCTAssertEqual(app!.user.y, 47.6)
+        XCTAssertEqual(app!.user.x, -122.4)
+        XCTAssertEqual(app!.shopDataFilename, "coffee_shop.csv")
```
At this point this test is not failing, it was written a little to late.

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -83,6 +83,24 @@ class CoffeeShopAppTest: XCTestCase {
+    }
+
+    func test_isNotFilename() {
+        let isFilename = CoffeeShopApp.isFilename("coffee_shop_csv")
```
I will argue that this is not needed, `coffee_shop_csv` is a perfectly valid file name, it does not need an extension.
I think that only thing needed here was that the filename was not empty.

eandreea replied:
Ok, I've over complicated things

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -32,6 +32,23 @@ class LocationTest: XCTestCase {
+    func test_coffeeShop() {
+        let coffeeShop = CoffeeShop(name: "Starbucks Seattle", location: Location(x: 47.5809, y: -122.3160))
+        XCTAssertEqual(coffeeShop.name, "Starbucks Seattle")
+        XCTAssertEqual(coffeeShop.location.x, 47.5809)
```
You retesting the `Location` constructor, you've already done this.

eandreea replied:
True, haven't thought of it this way. Should have compared the location property directly

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -32,6 +32,18 @@ class LocationTest: XCTestCase {
+
+// MARK: - CoffeeShop
+struct CoffeeShop {
+    var name: String
```
Again I think this structure should not be muttable.

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -166,6 +166,30 @@ class CoffeeShopTest: XCTestCase {
+            if distance1 > distance2 {
+                ordered = false
+            }
+        }
```
Is there any other way of comparing arrays in without doing this custom logic? One would not trust that the logic is accurate by glancing over tests.

eandreea replied:
I should have used a more specific test here

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -210,6 +214,31 @@ class CoffeeShopTest: XCTestCase {
+            let distance2 = orderedCoffeeShops[i+1].distance
+            if distance1 < distance2 {
+                ordered = false
+            }
```
This is duplicate as in test_coffeeShopsOrderedByDistanceToLocation

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -58,6 +58,39 @@ class CoffeeShopTest: XCTestCase {
         XCTAssertEqual(coffeeShop.location.y, -122.3160)
     }
+
+    func test_coffeeShopFromString() {
```
Same comment as before, way to many tests, the cycle should be way smaller, have a fail make it pass quickly

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -166,6 +166,30 @@ class CoffeeShopTest: XCTestCase {
+        XCTAssertNotNil(coffeeShops)
+        XCTAssertEqual(coffeeShops!.count, 6)
+        let location = Location(x: -122.4, y: 47.6)
+        let orderedCoffeeShops = CoffeeShop.sort(coffeeShops: coffeeShops, byDistanceTo: location, ascending: true)
```
Passing in booleans: https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -2,7 +2,7 @@ import Foundation
 
 // MARK: - Location
-struct Location {
+struct Location: Equatable {
```
You only make this structs Equatable because of the tests, this is usually a best practice, the production code should have meaning for production and should be driven by business requirements not implementation details.

eandreea replied:
Yes, I've added the conformance to the Equatable protocol only for testing purposes. I'l keep this in mind when refactoring

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -315,17 +328,65 @@ class CoffeeShopAppTest: XCTestCase {
+                        """
+
+        guard let coffeeShopsWithDistance = app!.closestCoffeeShops(count: 3, coffeeShopsCSV: fileContent) else {
+            NSLog("Invalid data format")
```
What is this used for?

eandreea replied:
Returns the closest 3 coffee shops from the coffee shops cvs file content string

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -315,17 +328,65 @@ class CoffeeShopAppTest: XCTestCase {
+
+    func test_3ClosestCoffeeShops() {
+        let app = CoffeeShopApp(["47.6", "-122.4", "coffee_shop.csv"])
+        XCTAssertNotNil(app)
```
Ref: http://defragdev.com/blog/?p=783

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -315,17 +328,65 @@ class CoffeeShopAppTest: XCTestCase {
+    NSLog("invalid paramenters for Coffee Shop application")
+}
 
+print("end")
```
What is this used for ?

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -166,6 +166,30 @@ class CoffeeShopTest: XCTestCase {
+        let location = Location(x: -122.4, y: 47.6)
+        let orderedCoffeeShops = CoffeeShop.sort(coffeeShops: coffeeShops, byDistanceTo: location, ascending: true)
+        var ordered = true
+        for i in 0..<coffeeShops!.count-1 {
```
This test is complicated.

It should be very easy to reason about a test, in this case I will simplify the preparation and only leave a couple of fake locations and make the assertion really easy and clear.

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -28,13 +28,22 @@ class LocationTest: XCTestCase {
         XCTAssertEqual(location.y, 1)
     }
+
+    func test_distanceToLocation() {
```
This is a good test, is small and clear. :+1:

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -28,13 +28,22 @@ class LocationTest: XCTestCase {
 // MARK: - CoffeeShop
 struct CoffeeShop: Equatable {
+
+    typealias CoffeeShopWithDistance = (coffeeShop: CoffeeShop, distance: Double)
```
A tiny bit of functionality slipped in :)

eandreea replied:
yes....so i don't have to write the tuple description every time and use the typealias instead

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -315,17 +328,65 @@ class CoffeeShopAppTest: XCTestCase {
+    let fileUrl = playgroundSharedDataDirectory.appendingPathComponent(app.shopDataFilename)
+    guard let fileContent = try? String(contentsOf: fileUrl) else {
+        NSLog("Unable to read \(app.shopDataFilename) from ~\\Documents\\Shared Playground Data")
+        break done
```
I don't think this is required, I would expect to throw a custom exception if the input is invalid instead of using this goto like statement.

eandreea replied:
By "Malformed entries should cause the program to exit appropriately" request I thought that the app should not throw and error, that's why I logged the messages in the console. I'll refactor these parts

alexandru-calinoiu replied:
`exit appropriately` can mean that you throw a explicit error if the user input is not right.

The best explanation I've found about this pattern is in the rust documentation file:///home/calin/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share/doc/rust/html/book/2018-edition/ch09-03-to-panic-or-not-to-panic.html, panic! is like throwing an exception.

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -0,0 +1,413 @@
+        self.init(name: components[0], location: Location(x: x, y: y))
+    }
+
+    static func coffeeShops(fromCSV content: String) -> [CoffeeShop]? {
```
This looks like a [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) can you refactor it to be more clear?

eandreea replied:
Yeah....not really what i wanted in the first place, but forgot to refactor it after correctly reading the file.

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -96,6 +63,125 @@ struct CoffeeShop: Equatable {
+
+// MARK: - unit tests
+
+// uncomment to run tests
```
Running tests should be straight forward and enforced or else they will be considered optional - tests are never optional (they prove that your code works).

Related to comments: 
https://refactoring.guru/smells/comments
https://blog.submain.com/field-guide-code-comments/

eandreea replied:
I've commented them because I couldn't run the playground from the terminal. But since I found out yesterday that this was not required, I'll do some refactoring

alexandru-calinoiu commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -0,0 +1,413 @@
+        return coffeeShops
+    }
+
+    static func sort(coffeeShops: [CoffeeShop], byDistanceTo location: Location, ascending: Bool = true) -> [CoffeeShopWithDistance] {
```
I don't believe this static function should belong to this class, it has very little to do with the class that is hosted into

eandreea replied:
This should have been refactored with the other static fct

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -0,0 +1,413 @@
+import Foundation
+
+// MARK: - Location
+struct Location: Equatable {
```
:+1:

alexandru-calinoiu reviewed (changes requested):
Overall:
- the use of static methods instead of having specialized services/classes makes code fell very procedural
- some unit tests try to do to much

Please review comments that have questions.

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -0,0 +1,413 @@
+}
+
+
+typealias CoffeeShopWithDistance = (coffeeShop: CoffeeShop, distance: Double)
```
How do you decide whether to use typealias or struct?

eandreea replied:
The struct is a coffee shop while the typealias is a renamed tuple which contains a coffee shop and a distance to the provided user location

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -0,0 +1,413 @@
+        self.location = location
+    }
+
+    init?(_ line: String, separator: String) {
```
What is the `_` used for here ?

eandreea replied:
It means that the label of the argument is missing 
https://docs.swift.org/swift-book/LanguageGuide/Functions.html#ID166

pirvudoru commented on CoffeeShop.playground/Contents.swift:
```diff
@@ -0,0 +1,413 @@
+}
+
+// MARK: - CoffeeShopApp
+struct CoffeeShopApp {
```
Prefer to have 1 class / file. It would be much easier to navigate.

eandreea replied:
I'll refactor this

pirvudoru reviewed (changes requested):
Hello. Thank you for submitting the solution. I'm not a swift developer. Please check the comments.

pirvudoru commented:
Also see: https://chris.beams.io/posts/git-commit/

eandreea committed:
Move LocationTest to single file, fail tests

eandreea committed:
Refactor Location, move to single file, pass tests

eandreea committed:
Move CoffeeShopTest to single file, fail tests

eandreea committed:
Refactor CoffeeShop, move to single file, pass tests

eandreea committed:
Fail new coffee shop from string with empty seprarator

eandreea committed:
Pass test coffee shop from string with empty separator

eandreea committed:
Fail new coffee shop from string with less data

eandreea committed:
Pass test coffee shop from string with less data

eandreea committed:
Fail test new coffee shop from string with more data

eandreea committed:
Pass test new coffee shop from string with more data

eandreea committed:
Fail test new coffee shop from string with empty name

eandreea committed:
Pass test new coffee shop from string with empty name

eandreea committed:
Fail test new coffee shop from string with invalid x

eandreea committed:
Pass test new coffee shop from string with invalid x

eandreea committed:
Fail test new coffee shop from string with invalid y

eandreea committed:
Pass test new coffee shop from string with invalid y

eandreea committed:
Move CoffeeShopAppTest to single file, fail tests

eandreea committed:
Refactor CoffeeShopApp, move to single file, pass tests

eandreea committed:
Fail test ordered coffee shops closest to user location

eandreea committed:
Pass test ordered coffee shops closest to user location

eandreea committed:
Fail test ordered coffee shops closest to user location when missing input file

eandreea committed:
Pass test ordered coffee shops closest to user location when missing input file

eandreea committed:
Fail test ordered coffee shops closest to user location when input file is corrupted

eandreea committed:
Pass test ordered coffee shops closest to user location when input file is corrupted

eandreea committed:
Fail test ordered coffee shops closest to user location for invalid file content

eandreea committed:
Pass test ordered coffee shops closest to user location for invalid file content

eandreea committed:
Find the 3 closest coffee shops

eandreea committed:
Fix coffee shops creation from string

MihaiPantiru commented on CoffeeShop.playground/Sources/CoffeeShopAppTest.swift:
```diff
@@ -0,0 +1,67 @@
+
+public class CoffeeShopAppTest: XCTestCase {
+
+    public override func setUp() {
```
@eandreea You should avoid overriding methods with no implementation. Is just "noise" to code base, no value added.

MihaiPantiru commented on CoffeeShop.playground/Sources/CoffeeShop.swift:
```diff
@@ -0,0 +1,33 @@
+        guard !separator.isEmpty,
+            compoments.count == 3,
+            !compoments[0].isEmpty,
+            let x = Double(compoments[2]),
```
@eandreea The results are wrong when running the app. This is happening because on read x is the last argument and y is the second one. After comparing with user location you get the results by comparing the user x with local y actually. Please use here x as second arg and y as the third argument.

eandreea replied:
Yes, as you can see, I've done this only in the last commit when I've noticed that the results were different from the ones before the refactoring. But this was request in the readme file.

MihaiPantiru commented on CoffeeShop.playground/Sources/LocationTest.swift:
```diff
@@ -0,0 +1,29 @@
+import Foundation
+import XCTest
+
+extension Location: Equatable {
```
@eandreea The extension is not used on this file.

eandreea replied:
You are right, I've haven't placed it correctly, haven't looked at this from this point of view

eandreea replied:
Since I'm using the conformance of Location to Equatable protocol in both CoffeeShopTest and CoffeeShopAppTest, and I have to put the code only in one place, I support my first decision to put it in LocationTest. If the equality test would have been required in only one location, then I would support that the code be moved in that file

MihaiPantiru replied:
What about having the Equatable conformance in a separated file if is needed in multiple places?

eandreea replied:
Of course that works too, but I don't consider it necessary here. If something more complex was required, then the separate file would be best.

MihaiPantiru commented on CoffeeShop.playground/Sources/CoffeeShopTest.swift:
```diff
@@ -0,0 +1,61 @@
+import Foundation
+import XCTest
+
+extension CoffeeShop: Equatable {
```
@eandreea As a good practice, is not necessary to extend existing functionality just to write tests. For example, instead of comparing coffeeShop!.location with  Location(x: 47.5809, y: -122.3160), you can compare  coffeeShop!.location.x with 47.5809

eandreea replied:
I agree, but in a previous comment it was stated that this functionality was already tested in the Location test, so it's just a duplicate. But to test that the CoffeeShop instance was correctly created I have to test the location parameter too... In the meantime i've seen that, after refactoring, only Location needs to conform to Equatable, so I'll remove the one from CoffeeShop

MihaiPantiru replied:
Great. Sorry for contradictory comments, we didn't sync the comments across the team. As long this is the result of another comment than is ok for me.

MihaiPantiru commented on CoffeeShop.playground/Sources/CoffeeShopTest.swift:
```diff
@@ -0,0 +1,61 @@
+        let coffeeShop = CoffeeShop(from: "Starbucks Seattle,-122.3160,47.5809", separator: ",")
+        XCTAssertNotNil(coffeeShop)
+        XCTAssertEqual(coffeeShop!.name, "Starbucks Seattle")
+        XCTAssertEqual(coffeeShop!.location, Location(x: 47.5809, y: -122.3160))
```
@eandreea You should avoid using force unwrappers. I can very easily make this code crash. Is better to have a guard and fail earlier. Ex
guard let sut = coffeeShop else {
 // fail
}
XCTAssertEqual(sut.name, "Starbucks Seattle")

eandreea replied:
Yeah, big mistake on my part, this was due to my forgetting to put continueAfterFailure = false in the test

MihaiPantiru reviewed (changes requested):
@eandreea Very good progress, the code is easier to read and better structured.
I still have some remarks, please read my comments.

eandreea committed:
Fix code based on latest comments

 - remove empty methods
 - remove unnecessary CoffeeShop conformance to Equatable
 - fix CoffeeShop test to not crash

alexandru-calinoiu commented:
Closing this as we will followup via email.
